### PR TITLE
Fix array size offset for inlined zero sized arrays in x86

### PR DESCRIPTION
**Pull request by bragaigor**
Fix array size offset for inline allocated zero sized arrays in x86 systems. The fix adds 2 extra instructions to the path of array allocations and 3 instructions in case of zero sized arrays.

Tests were conducted and the proposed fix does not impact throughput and performance

The fix adds 2 instructions to the critical path of array allocations, and 3 instructions in case of zero sized arrays. The 2 added instructions are a compare and a jump, and the third is a store to properly initialize the size field in case of zero sized array.

This is x86 assembly before the change:
```
############# Before change ############
…
rep stosb                        # SAME 1
mov rax, rdx                     # SAME 2
mov dword ptr [rax], 0x0010b300  # SAME 3
mov dword ptr [rax+0x4], r12d    # SAME 4
Label L0049:    
…             
```
And this is the x86 assembly after the change (Change labelled as # EXTRA), where EXTRA1/2 are the ones that would always be executed and # EXTRA 3 only executed in case of zero sized arrays:
```
############# After change #############
…
rep stosb                       # SAME 1
mov rax, rdx                    # SAME 2
mov dword ptr [rax], 0x0010b300 # SAME 3
cmp r12d, 0x00000000            # EXTRA 1
jne Label L0080                 # EXTRA 2
mov dword ptr [rax+0x8], r12d   # EXTRA 3* : Executed only if is zero sized array
Label L0080:
mov dword ptr [rax+0x4], r12d   # SAME 4
Label L0049:  
…
```

More detailed information can be found in the issue linked below.

Fixes: https://github.com/eclipse/openj9/issues/9683

Signed-off-by: Igor Braga <higorb1@gmail.com>

### CODE REVIEW

**Comment by 0xdaryl** [comment]
If you only care about a zero or non-zero result, your `cmp r12d, 0` should be replaced with a `test r12d, r12d`.  That should save one instruction byte in the worst case, or 2 bytes if a register other than r8 through r15 is chosen to hold the length.

**Comment by bragaigor** [comment]
Yes, you’re correct @0xdaryl , we only care about zero or non-zero results. I’ve included the changes and it still behaves as expected. Thanks!
Now we have:
```
…
rep stosb                       # SAME 1
mov rax, rdx                    # SAME 2
mov dword ptr [rax], 0x0010b300 # SAME 3
test r12d, r12d                 # EXTRA 1 <<<<<< updated
jne Label L0080                 # EXTRA 2
mov dword ptr [rax+0x8], r12d   # EXTRA 3* : Executed only if is zero sized array
Label L0080:
mov dword ptr [rax+0x4], r12d   # SAME 4
Label L0049:  
…
```

**Comment by andrewcraik** [comment]
Wouldn't it make more sense to do a cmov rather than branching around one instruction?

**Comment by 0xdaryl** [comment]
> Wouldn't it make more sense to do a cmov rather than branching around one instruction?

I think it's because `cmov` doesn't have a conditional store form, only loads.

**Comment by andrewcraik** [comment]
Ah true, then my next question - given the minimum object size is 16 bytes (I believe) when would it be incorrect to run the additional store? I know the mainline code may write a value to the location, but would it ever be wrong to write it given the offset is +0x8 it won't be outside the actual object's allocated size... just trying to think if there is a way to skip the branch around one instruction...

**Comment by bragaigor** [comment]
The additional store happens in case of zero sized arrays which have the layout: `class + mustBeZero + size + padding`. And given this path is only taken in case of array allocation, the minimum array size would be when the array has exactly 1 element, where we would have the layout: `class + size + 1st_elem`. In which case, even if we're dealing with char array, we would still be under the minimum object size (also because of alignment)? So if we unconditionally store at offset `+0x8` in that case we would be initializing that first element, which should be benign?

**Comment by bragaigor** [comment]
I double checked the array object layout and if we have a char array sized 1 where `sizeof(char) == 1` (1 byte) we would have the layout in compressed refs:
```
|--00--| |--11--| |--22--| |--33--|
00044c00 00000001 00000000 00000000
# Where
|--00--| == 00044c00 == class (32 bits)
|--11--| == 00000001 == size (32 bits)
|--22--| == 00000000 == first byte as char and rest as padding/alignment (32 bits)
|--33--| == 00000000 == padding/alignment (32 bits)
```
And if we have a zero sized array it looks like:
```
00044c00 00000000 00000000 00000000
# Where
|--00--| == 00044c00 == class (32 bits)
|--11--| == 00000000 == mustBeZero (32 bits)
|--22--| == 00000000 == size (32 bits)
|--33--| == 00000000 == padding/alignment (32 bits)
```

With Andrew's suggesting the path looks like the following, where we only have one additional instruction (for all array sizes in this case), in which case section `|--22--|` is always initialized to `0` regardless if it's a zero sized array or not:
```
…
rep stosb                             # SAME 1
mov rax, rdx                          # SAME 2
mov dword ptr [rax], 0x0010b300       # SAME 3
mov dword ptr [rax+0x4], r12d         # SAME 4
mov dword ptr [rax+0x8], 0x00000000   # EXTRA: Always executed <<<<<<<<<<<<<<<<<<<<<<
Label L0049:  
…
```
Is this prefered @andrewcraik ? If so should I run some workloads to check for any regression? From the tests conducted here #9683 it doesn't seem that one instruction will cause any penalties, but I could be wrong. I tested and this approach also fixes the problem and everything else still behaves as expected.

**Comment by andrewcraik** [comment]
Comparing your original sequence:
```
rep stosb                       # SAME 1
mov rax, rdx                    # SAME 2
mov dword ptr [rax], 0x0010b300 # SAME 3
test r12d, r12d                 # EXTRA 1 <<<<<< updated
jne Label L0080                 # EXTRA 2
mov dword ptr [rax+0x8], r12d   # EXTRA 3* : Executed only if is zero sized array
Label L0080:
mov dword ptr [rax+0x4], r12d   # SAME 4
Label L0049:  
...
```
I would say that the much shorter unconditional sequence would be better and faster if it is always correct:
```
…
rep stosb                             # SAME 1
mov rax, rdx                          # SAME 2
mov dword ptr [rax], 0x0010b300       # SAME 3
mov dword ptr [rax+0x4], r12d         # SAME 4
mov dword ptr [rax+0x8], 0x00000000   # EXTRA: Always executed <<<<<<<<<<<<<<<<<<<<<<
Label L0049:
...
```
my only additional question is whether the `rep stosb` could be adjusted to zero this piece of memory (eg zero from start of length through and then write the length rather than just zero the body)? That would save the extra mov potentially and speed this up a bit more?

**Comment by bragaigor** [comment]
Yes, the simple tests that I ran worked fine. I'll run further tests to make sure the second solution did not break anything else. And as for making `rep stosb` initialize the header it could work but if `canSkipZeroInitialization()` is set to True or if `initInfo` is available, then `rep stosb` is skipped which would skip the size initialization. So I'm not sure if `rep stosb` will always be called in case of array initialization?

**Comment by andrewcraik** [comment]
Might not be but if we are using it perhaps we do the zero with that and otherwise do the unconditional store? This is performance code so those kinds of things would be warranted to reduce the overhead if there is benefit.

**Comment by bragaigor** [comment]
Ah yes that logic make sense, let me look into it.

**Comment by bragaigor** [comment]
After further investigation it looks like the `size` field of zero sized arrays already gets initialized by `rep stosb` with the existing code if we use compressed refs. I observed that by walking instruction by instruction in gdb and the code also reflects that; `genZeroInitObject2()` does that by subtracting the size of a contiguous array header (which is always 8 bytes in compressed refs):
```
auto headerSize = node->getOpCodeValue() != TR::New ? TR::Compiler->om.contiguousArrayHeaderSizeInBytes() : TR::Compiler->om.objectHeaderSizeInBytes();
```
And initializes the other 8 bytes, which does include the `size` field from a zero sized array. So in the following representation of a zero sized array:
```
|--00--| |--11--| |--22--| |--33--|
00044c00 00000000 00000000 00000000
```
`rep stosb` initializes `|--22--|` and `|--33--|` so life is good. The problem only arises when `rep stosb` path is not executed, in other words, when `initInfo` is available or when `canSkipZeroInitialization()` is true. What is missing is a check in case we don't call `genZeroInitObject2()`. So I'll update the code to include this check and substitute `cmp` and `test` with the unconditional store. But not so fast... this only concerns the compressed refs part of the story. 

If we are using full refs, it looks like `rep stosb` becomes a `nop` because the size of contiguous and discontiguous array headers with full references are the same (16 bytes), unlike in compressed refs where the contiguous array headers is half the size of discontiguous array headers. Therefore this operation in `genZeroInitObject2()` which calculates the number of bytes to initialize by `rep stosb`:
```
generateRegImmInstruction(SUBRegImms(), node, tempReg, headerSize, cg);
```
results in `0` being stored in `tempReg` which is `rcx`; hence,`rep stosb` become a `nop`. Furthermore, `genInitArrayHeader` erroneously stores `size` in the wrong field when it comes to zero sized arrays, just like its compressed refs counterpart. I think in such case it would be best to do something similar to the compressed refs solution; where we add an unconditional store. To make it consistent with compressed refs we could modify `genZeroInitObject2()` to make `rep stosb` store the `size` field for us, which is probably the best solution? 

If we do that, this leaves us with a consistent situation where in both compressed and full references `rep stosb` takes care of initializing the `size` field. And in the case `rep stosb` path is not taken, we unconditionally generate a store for both compressed and full refs.

What you think @andrewcraik ?

**Comment by andrewcraik** [comment]
I think that sounds like a plan - perf verification of the different configurations changed will be required before we merge this, but making thins consistent and eliminating the extra store when it can be fused into the rep stosb makes sense to me and should help keep things simple. Please make sure to litter some comments into the code to outline why things are the way they are since it isn't obvious from just scanning through the code (an existing problem rather than one you created, but would be nice to see things improved on that front while you are in the area if it isn't a huge burden). Thanks for all your work on this an willingness to consider other sequences!

File: runtime/compiler/x/codegen/J9TreeEvaluator.cpp [hunk 1]
```diff
@@ -7840,6 +7873,7 @@ J9::X86::TreeEvaluator::VMnewEvaluator(
          if (canUseFastInlineAllocation)
             {
             useRepInstruction = genZeroInitObject2(node, objectSize, elementSize, sizeReg, targetReg, tempReg, segmentReg, scratchReg, cg);
+	    shouldInitZeroSizedArrayHeader = false;
```
Comment (bragaigor) [inline runtime/compiler/x/codegen/J9TreeEvaluator.cpp:7876]: Fix tab/spaces

**Comment by bragaigor** [comment]
I conducted some perf tests regarding this https://github.com/eclipse/openj9/pull/10946#issuecomment-714885191 solution. I’ve chosen workloads that has different proportions of zero sized arrays. Since the fix affects all GC policies, I tested against the most popular GC: gencon and balanced. I also color coded to better visualize the results, brown shades makes reference to gencon and blue shades makes reference to balanced. Master branch (dark shade) is a build WITHOUT the change, while Zero branch (light shade) contains the fix above. Furthermore, to exercise the different code paths of the introduced changes, I conducted tests in both compressed references mode (first results below) and non-compressed references mode:

**Compressed References Mode Results**

Machine1: 24 CPUs, 58GB RAM
SPECjvm2008 XML
Screen Shot 2020-10-29 at 1 17 33 PM

Machine1: 24 CPUs, 58GB RAM
SPECjbb2015
*Difference here mean <run_WITH_change_result> minus <run_WITHOUT_change_result>
```
# Averaged across 4 runs
Max JOPs mean difference:      -143. STD: ~330 (build WITHOUT changes had higher Max JOPs)
Critical JOPs mean difference: +231, STD: ~102 (build WITH changes had higher Critical JOPs)
```
Machine2: 28 CPUs, 98GB RAM
Daytrader
Screen Shot 2020-10-29 at 1 17 45 PM

**Non-compressed References Mode Results**

Machine1: 24 CPUs, 58GB RAM
SPECjvm2008 XML
Screen Shot 2020-10-29 at 1 18 31 PM

Machine1: 24 CPUs, 58GB RAM
SPECjbb2015
*Difference here mean <run_WITH_change_result> minus <run_WITHOUT_change_result>
```
# Averaged across 4 runs
Max JOPs mean difference:      +373. STD: ~570 (build WITH changes had higher Max JOPs)
Critical JOPs mean difference: +27, STD: ~245 (build WITH changes had higher Critical JOPs)
```

Machine2: 28 CPUs, 98GB RAM
Daytrader
Screen Shot 2020-10-29 at 1 18 43 PM
* Each row here corresponds to independent runs

I did not observe any regression on the tests conducted above. SPECjvm2008 and DayTrader throughput did not produce any discrepancies. And given the standard deviation for the SPECjbb2015 results, there doesn't seem to have any regressions there either. I'll have numbers for Daytrader + gencon and SPECjbb2015 + balanced next week. Are there other workloads/benchmarks and/or configurations that I should run for perf tests besides these ones?

**Comment by bragaigor** [comment]
Hello, here are the remaining numbers 

**Compressed References**

Machine4: 48CPUs 256GB RAM
SPECjbb2015
*Difference here mean <run_WITH_change_result> minus <run_WITHOUT_change_result>
```
# Averaged across 4 runs
Max JOPs mean difference:      -92. STD: ~520 (build WITHOUT changes had higher averaged Max JOPs)
Critical JOPs mean difference: -65, STD: ~455 (build WITHOUT changes had higher averaged Critical JOPs)
```

Machine2: 28 CPUs, 98GB RAM
Daytrader
daytrader_cmprss

**Non-compressed References**

Machine4: 48CPUs 256GB RAM
SPECjbb2015
*Difference here mean <run_WITH_change_result> minus <run_WITHOUT_change_result>
```
# Averaged across 4 runs
Max JOPs mean difference:      +435. STD: ~800 (build WITH changes had higher averaged Max JOPs)
Critical JOPs mean difference: +126, STD: ~195 (build WITH changes had higher averaged Critical JOPs)
```

Machine2: 28 CPUs, 98GB RAM
Daytrader
daytrader_full

Given standard deviation and throughput values there doesn't seem to be performance penalties here either. What you think @andrewcraik ?

**Comment by andrewcraik** [comment]
@bragaigor looks like it is a perf neutral change with those additional refinements you have made. I think this should be getting close to ready. Thank you for the collaboration on the sequence and for reworking things to reduce the perf overhead.

**Comment by andrewcraik** [comment]
Jenkins test sanity all jdk15

**Comment by andrewcraik** [comment]
@bragaigor you have completed functional testing to your satisfaction?

**Comment by bragaigor** [comment]
@andrewcraik that's correct, completed all correctness and performance tests to the best of my JIT knowledge

**Review by andrewcraik — APPROVED** [review]

**Comment by fjeremic** [comment]
Jenkins test sanity ppc64_aix,ppc64le_linux_xl,x86-64_linux,x86-64_linux_xl jdk15

**Comment by andrewcraik** [comment]
All green and perf tested - looks good to go. merging

**Comment by pshipton** [comment]
This is causing failures on x86 32-bit platforms (jdk8). Backing out https://github.com/eclipse/openj9/pull/10946/commits/37753b8f resolves them.
A number of the failures would have occurred in the OpenJ9 nightly build, but due do another change Windows 32-bit didn't compile. See internal build 459016 for the failures.

**Comment by pshipton** [comment]
I'm reverting this until the failure can be resolved.

**Pull request merged**